acstools/focus_diverse_epsfs.py

```diff
@@ -0,0 +1,460 @@
+    """Function to perform further spatial interpolations given the input ePSF array.
+    It uses bi-linear interpolation for the
+    integer pixel shifts, and bi-cubic interpolation for any specified subpixel phase shifts.
+
+    This function allows users to interpolate the provided ePSF arrays
+    to any arbitrary ``(x, y)`` coordinates. It can be called with ``pixel_space=True`` to
+    downsample the ePSF into detector space. Subpixel phase shifts can be applied by
+    setting ``subpixel_x`` and ``subpixel_y`` between 0 and 0.99.
+
+    .. note::
+
+        This function requires an optional ``scipy`` dependency to be installed
+        for ``pixel_space=True``.
+
+    .. note::
+
+        This function only works with 4096x4096 ACS/WFC frame.
+
+    Parameters
+    ----------
+    ePSFs : numpy.ndarray
+        Array with the ePSFs.
+
+    x : int
+        X-coordinate (0-indexed) of the desired output ePSF.
+
+    y : int
+        Y-coordinate (0-indexed) of the desired output ePSF. Please note that the range here is between
+        0 and 4095, inclusive; i.e., WFC1 runs from 2048-4095, inclusive.
+
+    pixel_space : bool
+        If `True`, downsample the ePSF into detector space.
+
+    subpixel_x, subpixel_y : float
+        The desired subpixel coordinate, between 0 and 0.99.
+
+    Returns
+    -------
+    P : numpy.ndarray or `None`
+        An ePSF array with the specified interpolation parameters, if successful.
+
+    Also see
+    --------
+    psf_retriever, multi_psf_retriever
+
+    """
+    valid_wfc_pixels = range(0, 4096)
+
+    if x not in valid_wfc_pixels:
+        LOG.error("The X coordinate should be an integer between 0 and 4096.")
+        return
+
+    if y not in valid_wfc_pixels:
+        LOG.error("The Y coordinate should be an integer between 0 and 4096.")
+        return
+
+    if subpixel_x < 0 or subpixel_x > 0.99 or subpixel_y < 0.0 or subpixel_y > 0.99:
+        LOG.error("Subpixel shifts should be between 0 and 0.99")
+        return
+
+    if not isinstance(pixel_space, bool):
+        LOG.error("pixel_space should be True or False.")
+        return
+
+    if not pixel_space and (subpixel_x != 0 or subpixel_y != 0):
+        LOG.error("Please set pixel_space=True to use the subpixel phase offsets.")
+        return
+
+    # round subpixel_x and subpixel_y to second decimal
+    subpixel_x = round(subpixel_x, 2)
+    subpixel_y = round(subpixel_y, 2)
+
+    # if user inputs coordinates for WFC1 (y >= 2048), then adjust down and set chip = "WFC1",
+    # otherwise assume chip = "WFC2"
+    chip = "WFC2"
+    if y >= 2048:
+        y = y - 2048
+        chip = "WFC1"
+
+    # FIXME: Confirm that these are actually correct in a 0-indexed system.
+    # give positions of ePSFs based on Andrea's coordinate system
+    acs_xCoords = np.array([0, 512, 1024, 1536, 2168, 2800, 3192, 3584, 4096])
+    acs_yCoords = np.array([0, 512, 1024, 1536, 2048])
```

Comment:
Hm, yeah, are these coordinates the centers of the ePSFs? Or the corner of each ePSF? (Seems impossible if they're the centers?)

Reply:
the PSFs are centered on the grid points shown in Figure 1 here: https://ui.adsabs.harvard.edu/abs/2018acs..rept....8B/abstract

extrapolation must have been used to generate those ePSFS

Reply:
I thought I just answered this but it looks like my answer disappeared? weird.

anyways, it seems some extrapolation was involved to generate the ePSFs at the grid edges.